File: dcc-credit-android/app/src/main/java/io/wexchain/android/dcc/base/ActivityCollector.java
package io.wexchain.android.dcc.base;

import android.app.Activity;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Created by Wangpeng on 2016/5/19.
 */
public class ActivityCollector {
    public static List<Activity> activities = new ArrayList<>();
    
    public static void addActivity(Activity activity) {
        activities.add(activity);
    }
    
    public static void removeActivity(Activity activity) {
        activities.remove(activity);
    }
    
    public static void finishAll() {
        for (Activity activity : activities) {
            if (!activity.isFinishing()) {
                activity.finish();
            }
        }
    }
    
    public static void finishActivity(Class tClass) {
        for (Activity activity : activities) {
            if (activity.getClass().equals(tClass)) {
                activity.finish();
            }
        }
    }
    
    public static void finishActivitys(List<Class> tClass) {
        if (tClass == null) {
            return;
        }
        for (Activity activity : activities) {
            for (Class c : tClass) {
                if (activity.getClass().equals(c)) {
                    activity.finish();
                }
            }
        }
    }
    
}

File: dcc-credit-android/app/src/main/java/io/wexchain/android/dcc/base/ActivityCollector.kt
package io.wexchain.android.dcc.base

import android.app.Activity
import io.reactivex.rxkotlin.subscribeBy
import io.reactivex.rxkotlin.toObservable
import java.util.*

/**
 * @author Created by Wangpeng on 2016/5/19.
 */
object ActivityCollector {

    var activities: MutableList<Activity> = ArrayList()

    val currentActivity: Activity
        get() = activities[activities.size - 1]

    fun getTaskActivity(index: Int): Activity {
        return activities[activities.size - 1 - index]
    }

    fun addActivity(activity: Activity) {
        activities.add(activity)
    }

    fun removeActivity(activity: Activity) {
        activities.remove(activity)
    }

    fun finishAll() {
        activities.filter {
            !it.isFinishing
        }.forEach {
            it.finish()
        }
    }

    fun finishActivity(tClass: Class<*>) {
        activities.filter { it.javaClass == tClass }.forEach { it.finish() }
    }

    fun finishActivitys(vararg tClass: Class<*>) {
        tClass.toObservable().filter {
            tClass.isNotEmpty()
        }.map { t ->
            activities.toObservable().filter {
                t == it.javaClass
            }
        }.subscribeBy {
            it.blockingForEach {
                it.finish()
            }
        }
    }

}
